singletonclient/SingletonClient/Implementation/SingletonConfig.cs

```diff
@@ -164,27 +299,313 @@ public string ReadResourceText(string resourceName)
+            {
+                return null;
+            }
+            if (locale == null)
+            {
+                locale = ConfigConst.DefaultLocale;
+            }
+
             string baseName = _resourceBaseName;
-            if (array.Length > 1)
+            string[] parts = _resourceBaseName.Split(new char[] { '.' });
+            resourceName = parts[0] + "." + resourceName;
+
+            if (ConfigConst.FormatResx.Equals(format))
             {
-                string[] parts = _resourceBaseName.Split(new char[] { '.' });
-                parts[parts.Length - 1] = array[1];
-                baseName = string.Join(".", parts);
-                resourceName = array[0];
+                return SingletonUtil.ReadResourceMap(resourceName, locale, _resourceAssembly);
             }
 
-            if (string.IsNullOrEmpty(resourceName))
+            return null;
+        }
+    }
+
+    public class SingletonConfigWrapper: ISingletonConfig
+    {
+        private IConfig _config;
+
+        private string _onlineUrl;
+        private string _offlineUrl;
+        private string _serviceUrl;
+
+        private string _product;
+        private string _version;
+
+        private string _bundleRoot;
+        private List<string> _externalComponentList;
+
+        private bool _isProductMode;
+        private bool _isOnlineSupported;
+        private bool _isOfflineSupported;
+        private bool _isSourceDefault;
+        private bool _isLoadOnStartup;
+
+        public SingletonConfigWrapper(IConfig config)
+        {
+            _config = config;
+
+            _onlineUrl = GetTextWithDefault(ConfigConst.KeyOnlineUrl, null);
+            _offlineUrl = GetTextWithDefault(ConfigConst.KeyOfflineUrl, null);
+
+            if (_onlineUrl != null)
             {
-                return SingletonUtil.ReadResourceMap(baseName, _resourceAssembly);
+                string[] strings = (_onlineUrl + "/").Replace("//", "/").Split('/');
+                _serviceUrl = strings[0] + "//" + strings[1];
+                ExtractProductInfo(strings);
             }
+            if (_offlineUrl != null)
+            {
+                _bundleRoot = _offlineUrl.Replace("file:///", "").Replace("://", "\x01");
+                _bundleRoot = (_bundleRoot + "/").Replace("//", "/");
+                string[] strings = _bundleRoot.Split('/');
+                ExtractProductInfo(strings);
 
-            Byte[] bytes = SingletonUtil.ReadResource(baseName, _resourceAssembly, resourceName);
-            string text = SingletonUtil.ConvertToText(bytes);
-            Hashtable bundle = parser.Parse(text);
-            return bundle;
+                BuildExternalComponentList();
+                _bundleRoot = _bundleRoot.Replace("\x01", "://");
+            }
+
+            IConfigItem configItem = _config.GetItem(ConfigConst.KeyProductMode);
+            _isProductMode = (configItem == null) ? true : configItem.GetBool();
```

Comment:
The feature of product mode is not finalized, you might need to remove it in future.